File: leetcode/50_pow.cpp
#include <iostream>
using namespace std;

class Solution {
public:
    double myPow(double x, int n) {

        if (abs(abs(x) - 1.0) <= 0.0000000000000001) {
            if (x > 0) {
                return 1.0;
            } else {
                if (n % 2 == 0) {
                    return 1.0;
                } else {
                    return -1.0;
                }
            }
        }

        int N = n/2;
        bool flag = false;
        double x_n = 1.0;
        if (N <= 0) {
            N *= -1;
            flag = true;
        }
        
        for (int i=0; i<N; i++) {
            if (flag) {
                x_n *= (1/x);
            } else {
                x_n *= x;
            }
        }

        if ((n % 2 == 1) || (n % 2 == -1)) {
            if (flag) {
                x_n *= (x_n * (1/x));
                
            } else {
                x_n *= (x_n * x);
            }
        } else {
                x_n *= x_n;
        }

        return x_n;
    }
};

int main(int argc, char* argv[]) {

    Solution s;
    double d = s.myPow(0.44894, -5);


    cout << "d = " << d << endl;

    return 0;
}

File: leetcode/50_pow.rs

struct Solution{}

impl Solution {
    pub fn my_pow(x: f64, n: i32) -> f64 {
        if n == 0 {
            return 1.0;
        }

        let mut res = 1.0;
        let mut x = x;
        let mut n = n;

        if n < 0 {
            x = 1.0 / x;
            if (n == -2147483648) {
                n = 2147483647;
                res *= x;
            } else {
                n = -n;
            }

            // n = -n;
        }

        while n > 0 {
            if n & 1 == 1 {
                res *= x;
                println!("res: {}", res);
            }
            x *= x;
            n >>= 1;
            println!(" n: {}", n);
        }

        return res;    
    }
}

fn main() {

    let x = 2.0;
    // let n = 7;
    let n = -2147483648;
    let res = Solution::my_pow(x, n);
    println!("{}", res);

}
